onnxruntime/contrib_ops/cuda/bert/attention_impl.cu

```diff
@@ -334,55 +334,132 @@ Status PrepareQkv(contrib::AttentionParameters& parameters,
+      }
+      qkv_format = AttentionQkvFormat::Q_K_V_BNSH;
     }
-    // cross attention with present state or self attention with present state
-    else if (data.past_key == nullptr && data.present_key != nullptr) {
-      assert(data.past_value == nullptr);
-      assert(data.present_value != nullptr);
-      assert(data.query != nullptr);
-      assert(data.key != nullptr);
-      assert(data.value != nullptr);
-
-      // TODO: supporting packed qkv for self attention may benefit performance
-      ORT_RETURN_IF_ERROR(LaunchTransQkv(stream, 1, sequence_length, batch_size, qk_head_size, num_heads,
-                          max_threads_per_block, false, data.query, q));
-
-      // TODO: supporting packed kv for cross attention may benefit performance
+#if USE_FLASH_ATTENTION
+    else if (use_memory_efficient_attention && data.past_key != nullptr && data.past_value != nullptr && parameters.pass_past_in_kv) {
+      // Transpose past_key and past_value to use memory efficient attention
+
+      // past_key (BxNxSxH) => temp_k_workspace (BxSxNxH)
+      ORT_RETURN_IF_ERROR(LaunchTransCtx(stream, kv_sequence_length, batch_size, qk_head_size, num_heads,
+                                         max_threads_per_block, false, data.past_key, data.temp_k_workspace));
+      // past_value (BxNxSxH_v) => temp_v_workspace (BxSxNxH_v)
+      ORT_RETURN_IF_ERROR(LaunchTransCtx(stream, kv_sequence_length, batch_size, qk_head_size, num_heads,
+                                         max_threads_per_block, false, data.past_value, data.temp_v_workspace));
+
+      // query => q, temp_k_workspace => k, temp_v_workspace => v
+      LaunchAddBias(stream, max_threads_per_block,
+              batch_size, sequence_length, kv_sequence_length,
+              num_heads, qk_head_size, v_head_size,
+              data.bias, data.query, data.temp_k_workspace, data.temp_v_workspace, q, k, v);
+
+      DUMP_TENSOR_D("q(BSNH)", q, batch_size * sequence_length, num_heads, qk_head_size);
+      DUMP_TENSOR_D("k(BSNH)", k, batch_size * kv_sequence_length, num_heads, qk_head_size);
+      DUMP_TENSOR_D("v(BSNH)", v, batch_size * kv_sequence_length, num_heads, v_head_size);
+      qkv_format = AttentionQkvFormat::Q_K_V_BSNH;
+
+      data.past_key = nullptr;
+      data.past_value = nullptr;
+    }
+    else if (use_memory_efficient_attention && data.present_key != nullptr && data.present_value != nullptr) {
+      // Use memory efficient attention kernel
+      LaunchAddBias(stream, max_threads_per_block,
+                    batch_size, sequence_length, kv_sequence_length,
+                    num_heads, qk_head_size, v_head_size,
+                    data.bias, data.query, data.key, data.value, q, data.temp_k_workspace, data.temp_v_workspace);
+
+      // temp_k_workspace (BxSxNxH) => present_k (BxNxSxH)
       ORT_RETURN_IF_ERROR(LaunchTransQkv(stream, 1, kv_sequence_length, batch_size, qk_head_size, num_heads,
```

Comment:
if using memory efficient attention in decoderinit, do we need to transpose the present since they serve as past in decoder next iteration which requires the shape as BSNH?

Reply:
The current exported version of Whisper expects the past inputs and present outputs to be of shape BNSH. If this is changed to shape BSNH, then the transpose is not needed.